### PR TITLE
Rewrite the Telestion Client CLI

### LINKED ISSUE #1377: Move from CommonJS to ModuleJS in `telestion-client-cli`
As of EOL for Node.js v12 we should switch to ModuleJS as script file format.

More information here:

- https://nodejs.org/api/modules.html

If you need a working example, please ask @fussel178.

### LINKED ISSUE #1378: Use Parcel as PSC builder in `telestion-client-cli`
To leave the webpack universe completely, we should update the build pipeline for PSCs in the `telestion-client-cli`, too.

Parcel provides a nice programmatic API: https://parceljs.org/features/parcel-api/

### PR DESCRIPTION
**Pull request by fussel178**
### Summary 

Rewrite the `tc-cli` to provide a better user experience, more functionality and better stability.

### Details 

This PR introduces a complete rewrite of the Telestion Client CLI package.
It features:
- new init command
- new build command based on Parcel and Electron Packager
- new start command with Hot Reloading, even in the Electron main process
- the source code is written in ModuleJS (CommonJS is deprecated since NodeJS v12 is EOL)
- better progress output
- better error messages
- programmatic API (when you want to write your CLI yourself)
- improved performance due switch to Parcel as web-app builder
- the CLI now requires pnpm as package manager (gets installed automatically)
- workspace link support (if you want to develop in the Telestion Client libraries and see your changes directly in a test project)

#### Preparation

You need the following tools:

- [NodeJS](https://nodejs.org/)
- [npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)
- [pnpm](https://pnpm.io/)

First, clone this repository and checkout the target branch:

```shell
git clone git@github.com:wuespace/telestion-client.git
cd telestion-client
git checkout feat/cli/module-js
```

Next, install the development dependencies via pnpm:

```shell
pnpm install
```

Build all packages:

```shell
pnpm build
```

#### Setup a test project with workspace link

First, uninstall any other versions of the Telestion Client CLI:

```shell
pnpm rm --global @wuespace/telestion-client-cli
npm rm --global @wuespace/telestion-client-cli
```

Go into the `@wuespace/telestion-client-cli` package and link it globally via pnpm:

```shell
cd packages/telestion-client-cli
pnpm link --global
```

Next go into a temporary directory and initialize a new Telestion Frontend Project:

```shell
cd ~/tmp
tc-cli init "Test Project 1 Workspace"
```

Wait for the initialization to finish.

Next, go into the created project and start the development process:

```shell
cd test-project-1-workspace
pnpm start
```

This starts Parcel in serve mode and configures Electron to execute the main process file and shows the Parcel development server.

Now, head back to the Telestion Client Monorepo and start the watch mode:

```shell
cd /path/to/telestion-client
pnpm watch
```

Now, you can edit code in the monorepo and the test project should automatically detect the changes and should rerender.

> Note: Due to limitations in Parcel, the development server in the test project doesn't detect your changes in the monorepo.
> A GitHub issue tracks the current progess: https://github.com/parcel-bundler/parcel/issues/4332

#### Setup a test project via private registry

First, uninstall any other versions of the Telestion Client CLI:

```shell
pnpm rm --global @wuespace/telestion-client-cli
npm rm --global @wuespace/telestion-client-cli
```

Next, install [Verdaccio](https://verdaccio.org/) and start it:

```shell
pnpm add --global verdaccio
verdaccio
```

Next, setup pnpm to use the current Verdaccio instance as the main npm registry:

```shell
pnpm set registry http://localhost:4873/
pnpm adduser --registry http://localhost:4873/
```

Now, go into the Telestion Client monorepo and increase the package versions.

You can use `sed` to automate this step:

```shell
sed -i 's/"version": "0.18.1"/"version": "0.19.0"/g' package.json packages/*/package.json
```

Now, publish the bumped packages to Verdaccio:

```shell
pnpm publish --no-git-checks --recursive --filter './packages/**/*'
```

Next go into a temporary directory and initialize a new Telestion Frontend Project:

```shell
cd ~/tmp
tc-cli init "Test Project 2"
```

Wait for the initialization to finish.

Next, go into the created project and start the development process:

```shell
cd test-project-2
pnpm start
```

You can also build the test project:

```shell
pnpm build
```

#### Publish changes to the private registry

If you changes in the monorepo that you want to the test in the test project, you need to publish them in your private registry.

First, rebuild all packages:

```shell
cd /path/to/telestion-client
pnpm build
```

Bump the package versions of the packages you've changed:

```shell
sed -i 's/"version": "0.19.0"/"version": "0.19.1"/g' packages/telestion-client-{common,cli}/package.json
```

And publish them:

```shell
pnpm publish --no-git-checks --recursive --filter './packages/**/*'
```

Head over to the test project and install the new versions:

```shell
pnpm add @wuespace/telestion-client-common@latest
pnpm add -D @wuespace/telestion-client-cli@latest
```

That's it!

### Related links 

- Fixes #1377 
- Fixes #1378 

### CLA

- [x] I have signed the individual contributor's license agreement **and sent** it to the board of the WüSpace e. V. organization.

### CODE REVIEW

File: packages/telestion-client-cli/src/model/index.mts [hunk 1]
```diff
@@ -0,0 +1 @@
+export * from './commands.mjs';
```
Comment (pklaschka) [inline packages/telestion-client-cli/src/model/index.mts:1]: Potentially missing an export from `'./parcel-events.mts'`?

**Review by pklaschka — CHANGES_REQUESTED** [review]
# Partial Review

This review is limited to reviewing the following scopes

- Overall review of functionality
- Code review in all packages but the `telestion-client-cli`
- Code review of the root files in the `telestion-client-cli` package
- Code review of the `api.mts` and `cli.mts` in the `telestion-client-cli` package
- Code review of the `actions/*.mts` files in the `telestion-client-cli` package
- Code review of the `model/*.mts` files in the `telestion-client-cli` package

## Overall
Very nice changes. We will have to coordinate how to update the documentation in time for this update, but other than that (and a few change requests), this looks awesome! :tada:

## Smaller change requests
- missing comments here and there (see inline comments)
- readability improvements (see inline comments)
- Promise performance improvements
- Are the increased version numbers intentional?

## Missing (Changes requested)
- Plugin system
- `tc-cli install` - a wrapper around dependency installation to simplify installing `electronDependencies`

File: packages/parcel-reporter-tc-cli/src/reporter.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import { Reporter } from '@parcel/plugin';
+import { JsonSerializable } from '@wuespace/telestion-client-types';
+
+export default new Reporter({
```
[the remark this answers is not in the document]
Comment (fussel178) [reply packages/parcel-reporter-tc-cli/src/reporter.ts:17]: Done in [`4a8dde4` (#1396)](https://github.com/wuespace/telestion-client/pull/1396/commits/4a8dde476d870668ac52341394c62b11c6f96600)

File: packages/parcel-resolver-react/src/resolver.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import { Resolver } from '@parcel/plugin';
+import { join } from 'path';
+
+export default new Resolver({
```
[the remark this answers is not in the document]
Comment (fussel178) [reply packages/parcel-resolver-react/src/resolver.ts:12]: Done in [`8afae17` (#1396)](https://github.com/wuespace/telestion-client/pull/1396/commits/8afae17cc8eb49ee7a05cf4038da8c411f085851)

File: packages/telestion-client-cli/src/actions/template.mts [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+ * @param electronDependencies - an array that contains the Electron dependencies for the new project
+ */
+export async function getTemplateReplacers(
+	prettyName: string,
+	moduleName: string,
+	dependencies: Record<string, string>,
+	devDependencies: Record<string, string>,
+	electronDependencies: string[]
+): Promise<TemplateReplacers> {
+	const addIndentation = (stringified: string) =>
```
[the remark this answers is not in the document]
Comment (fussel178) [reply packages/telestion-client-cli/src/actions/template.mts:196]: Done in [`fa75a24` (#1396)](https://github.com/wuespace/telestion-client/pull/1396/commits/fa75a244361e043ea811f50c417d0bf6bbaeb373)

File: packages/telestion-client-cli/src/actions/workspace.mts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+const logger = getLogger('Workspace Action');
+
+const tagFileName = 'workspace.txt';
+
+const tagFileContent =
+	"!!! DON'T EDIT OR REMOVE !!!\n\n" +
+	'This project contains workspace dependencies.\n' +
+	"Some features like native dependencies aren't available.\n" +
+	'If you want to use the full feature set, install @wuespace/telestion-client-cli ' +
+	'from the NPM registry and initialize a new project.\n';
```
[the remark this answers is not in the document]
Comment (fussel178) [reply packages/telestion-client-cli/src/actions/workspace.mts]: Done in [`bc61b0e` (#1396)](https://github.com/wuespace/telestion-client/pull/1396/commits/bc61b0e3e4aae156661b337b63b8ea7469b1872c)

File: packages/telestion-client-cli/src/model/index.mts (hunk 1 above)
Comment (fussel178) [reply packages/telestion-client-cli/src/model/index.mts:1]: Done in [`9c15928` (#1396)](https://github.com/wuespace/telestion-client/pull/1396/commits/9c15928bbbf0d1f30a7c72ca7d7b081c3d878fb6)

**Comment by fussel178** [comment]
>     * Plugin system

Ithink, we've already discussed this topic. From my point of view, t's okay that we merge this PR and then add an additional package to the monorepo that controls and manages the Electron main process. (with access hooks, properties, etc.)
 
>     * `tc-cli install` - a wrapper around dependency installation to simplify installing `electronDependencies`

Good point! I'll implement this feature before merging this PR. :wink:

**Review by pklaschka — APPROVED** [review]
LGTM. Thank you for your contribution 🎉 🙂

**Comment by pklaschka** [comment]
> Ithink, we've already discussed this topic. From my point of view, t's okay that we merge this PR and then add an additional package to the monorepo that controls and manages the Electron main process. (with access hooks, properties, etc.)

Agreed, but then, we should definitely add a braking change notice about plugins in the meantime

**Comment by fussel178** [comment]
> Agreed, but then, we should definitely add a braking change notice about plugins in the meantime

Good point. Hmm, currently we build the PSC only in an unpacked format (`linux-x64-unpacked`, `mas-arm64-unpacked`). The package options (like `.rpm`, `.deb`, `.exe`, `.app`, `.msi`, etc.) are currently not working. This feature is essential for the build pipeline. I think, we also add it via a separate PR.

**Review by jantischhoefer — APPROVED** [review]
Great work Ludwig, i don't have anything to complain about.

**Pull request merged**